library/core/src/option.rs

```diff
@@ -888,6 +902,13 @@ impl<T> Option<T> {
                 // such that this will be in-bounds of the object.
                 unsafe { self_ptr.byte_offset(Self::SOME_BYTE_OFFSET_GUESS).cast() }
             };
+        #[cfg(not(bootstrap))]
+        let payload_ptr: *const T =
+        // SAFETY: The intrinsic only requires a const pointer to a valid
+        // `Option and creates an in-bounds pointer into the space where the
+        // `Some(_)` value potentially is.
+            unsafe { crate::intrinsics::option_payload_ptr(self as *const Option<T>) };
```

Comment:
As I mentioned in <https://github.com/rust-lang/rust/pull/109095#discussion_r1136255442>, please don't use `as` for this.  IIRC `LHS as *const` on a `&mut` is `(&*LHS) as *const`, which would mean that the pointer would have a `SharedReadOnly` tag, rather than the `Unique` tag that we want it to have.

```suggestion
            unsafe { crate::intrinsics::option_payload_ptr(crate::ptr::from_mut(self).cast_const()) };
```

Reply:
Yup, confirmed that `as` is `&raw const` instead of `&raw mut`: <https://rust.godbolt.org/z/efq1zoM8o>.

Reply:
Good catch! I haven't worked with `cast_const()` or `cast_mut()` before, but they're certainly nice.